KEP 831-Kubeflow-Helm-Support: Support Helm as an Alternative for Kustomize (#832)
kubeflow/community: proposals/831-helm-manifests/README.MD

> # 831-Helm-Manifests: Installing Kubeflow with Helm

Reviewer: This KEP should be renamed:
```suggestion
# 831-Helm-Manifests: Support Helm for Individual Kubeflow Tools
```

Author: But is should also allow users to combine the Helm charts on their own as needed downstream. Therefore at least the title i would leave as it is

## Summary

> Kubeflow manifests provide a fast way to deploy a minimal Kubeflow platform, with best-effort community support. For guaranteed assistance, users can opt for [commercial support](https://www.kubeflow.org/docs/started/support/). This approach extends to Helm support. Contributions and bug reports are encouraged, but no support will be guaranteed. The goal is to build a similar folder structure as we have for the current Kustomize manifests to rely on the existing CI/CD and have a signle source of truth. We also want to incorporate best-practices from Argo for Kubeflow Helm manifests.

Reviewer: Please update this summary to tell that we want to add support Helm Charts for each Kubeflow tool.

Author: done

## Motivation

The demand for a Helm chart for a basic Kubeflow installation has increased. Given the KSC's stance in issue 821 on neutral deployment language and user-defined production readiness, this is an opportune time to introduce a Helm chart. Supporting Helm will enhance ease of adoption and simplify deployments while maintaining the flexibility of community-maintained manifests. There have already been community efforts as well as the Kubeflow-Helm-Chart Slack Channel.

Currently, because Kubeflow/manifests only contains Kustomize but no Helm manifests, many potential users and companies that require Helm manifests due to company processes/policies have to rely on third-party options. While these options are valuable, they require engagement with adjacent projects and communities and often common improvents done in one fork do not reach and benefit the whole community.

> As a project, we must ensure that our Helm chart provides a quick and accessible way for users to deploy a complete Kubeflow platform and individual components, enabling them to manage their environments or adopt a vendor solution.

Reviewer: This needs to be removed.

```suggestion
```

> Simplifying the Kubeflow deployment and customization lowers the barrier to entry, increases adoption, and encourages contributions. Just as Kubernetes enabled a new wave of cloud-native startups, a neutral, accessible deployment path can empower AI/ML startups to leverage tools like the Training Operator or Katib without reinventing common maintenance patterns. If support becomes burdensome, teams can obtain [commercial support](https://www.kubeflow.org/docs/started/support/).

Reviewer: ```suggestion
Simplifying the deployment of Kubeflow Tools and customization lowers the barrier to entry, increases adoption, and encourages contributions. Just as Kubernetes enabled a new wave of cloud-native startups, a neutral, accessible deployment path can empower AI/ML startups to leverage tools like the Kubeflow Trainer or Katib without reinventing common maintenance patterns. If support becomes burdensome, teams can obtain [commercial support](https://www.kubeflow.org/docs/started/support/).
```

By making deployment easy, we attract more end users and foster collaboration with broader communities like PyTorch, improving our implementations in service of their users.

### About Helm

- Helm is a graduated project in the CNCF and is maintained by the Helm community.
- Helm is supported by and built with a community of over 400 developers.
- Helm helps you manage Kubernetes applications, rollbacks, updates, dependencies, and releases, and the most complex Kubernetes applications. [More about Helm](https://helm.sh/)

## Value to the Community

- Streamline the deployment, upgrade, and rollback of the kubeflow installation process

> - Provide another way to install Kubeflow to give the community more options, promoting flexibility and choice

Reviewer: ```suggestion
- Provide another way to install Kubeflow Tools to give the community more options, promoting flexibility and choice
```

## Goals

> ✅ A fully functional Kubeflow Helm chart for the targeted release. Users can install Kubeflow and subsets as a multi-tenant platform and also the individual projects as single-tenant without platform-level support.
> ✅ The kustomize Manifests are still the single source of truth for the Helm manifests and they shall use the same CI/CD to make sure that they satisyfy the same requirements and the difference is negligible. Therefore they must be next to each other in the same repository for Kstomize and Helm manifests. After the initial PoC the Helm manifests should be upstreamed to the respective working groups and be maintained there and synchronized like the kustomize manifests.

Reviewer: This should be changed to Helm Charts for Individual Kubeflow Tools
```suggestion
✅ Provide Helm Charts for Individual Kubeflow Tools
```

✅ Published Helm chart documentation with straightforward and uncomplicated configuration options.
✅ Contribution to the Kubeflow community effort for Helm-based installation as part of the official Kubeflow repository.   
✅ Where possible, rely on upstream Helm charts (i.e., Istio) instead of translating our own Kustomize manifests again.
✅ Consolidate community Helm efforts and prevent duplicate efforts.

> ## Non-Goals

Author: No that is exactly what should be doable downstream by users. Even if we do not provide it by default it should not be prevented.

**Mitigation:**
* Adopt a community-driven maintenance model, similar to Argo’s Helm charts.
* Establish clear ownership within the Kubeflow community and define a process for versioning and deprecating charts.
* Regularly sync Helm charts with upstream manifests to prevent drift.

3. Security Considerations

**Risk:** Misconfigured Helm deployments could introduce security vulnerabilities, such as exposed services, weak authentication, or misconfigured role-based access control (RBAC).

**Mitigation:**
* Follow Kubernetes security best practices, ensuring charts include secure default configurations.
* Conduct security reviews as part of Kubeflow’s release cycle.
* Provide Helm values presets for secure and production-ready configurations.

## Design Details

### Helm Chart Structure

The repository will contain a root Helm chart (kubeflow) that acts as an umbrella for subcharts: 

```
kubeflow/manifests/experimental/helm
│── charts/
│   │── training-operator/
│   │── katib/
│   │── pipelines/
│   │── istio/
│   │── profiles/
│   │── common/
│   │── kserve/
│── templates/
│── values.yaml
│── Chart.yaml

> │── README.md

Author: maybe @kunal-511 can do so.

     size: 10Gi

katib:
 enabled: false

training-operator:
 enabled: true
 resources:
   limits:
     cpu: "2"
     memory: "4Gi"
```

### Installation

To install Kubeflow Pipelines and the Training Operator only: 

```
helm install kubeflow ./kubeflow-helm-chart --set pipelines.enabled=true --set training-operator.enabled=true
```

To enable Katib after the initial installation: 

```
helm upgrade kubeflow ./kubeflow-helm-chart --set katib.enabled=true
```

To roll back a deployment: 

```
helm rollback kubeflow 1
```

### Security and Default Configurations

To ensure secure and production-ready deployments, the Helm chart will include:

* Minimal privileges using Role-Based Access Control (RBAC) and Pod Security Standards restricted.
* Network policies to restrict component communication where necessary.
* Secure default values, with optional overrides for users needing customization.

Example RBAC template (templates/rbac.yaml):

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
 name: kubeflow-pipelines-role
rules:
- apiGroups: ["kubeflow.org"]
 resources: ["pipelines"]
 verbs: ["get", "list", "watch"]
```

### Implementation Plan

> * Create a subdirectory (kubeflow/manifests/experimental/helm) to host the Helm charts.

Author: Because you want someone to mentor this. I do not have the time to deal with each maintainer and fix their tests, i need the extended platform tests and make sure that the rendered output is identical with kustomize. All of this must be reusable and scalable without code duplication. So i can incubate in /experimental with full proper tests support and then upstream. Otherwise i would rather cancel the whole helm effort and move to something time efficient, because i value my time and that is not worth it.

> * Create a subdirectory (kubeflow/manifests/experimental/helm) to host the Helm charts.

Reviewer: In that case, we should have plan how we can keep those Helm Charts synchronized between `kubeflow/manifests` and other repositories.

This this KEP is affecting every Kubeflow project, so it requires collaboration across working groups and project maintainers.

* Define the Helm chart structure, ensuring compatibility, synchronizability, and a single source of truth with the Kustomize manifests.

> * Develop subcharts for each major Kubeflow component (Pipelines, Platform, Notebooks, Dashboard, Katib, Training Operator, Istio, etc.) and make them deployable simultaneously to replicate the kustomize manifests.

Reviewer 2: How much time have we spent on this? To block a community-supported Helm chart? Fascinating.

> * Develop subcharts for each major Kubeflow component (Pipelines, Platform, Notebooks, Dashboard, Katib, Training Operator, Istio, etc.) and make them deployable simultaneously to replicate the kustomize manifests.

Reviewer 3: I 100% agree with @chasecadet here.

> * Develop subcharts for each major Kubeflow component (Pipelines, Platform, Notebooks, Dashboard, Katib, Training Operator, Istio, etc.) and make them deployable simultaneously to replicate the kustomize manifests.

Reviewer 2: Let them eat cake.

> * Develop subcharts for each major Kubeflow component (Pipelines, Platform, Notebooks, Dashboard, Katib, Training Operator, Istio, etc.) and make them deployable simultaneously to replicate the kustomize manifests.

Author: Yes, but we explicitly did not block Istio or knative or dependencies in general. Helm becomes useful for large installations, where you need templating for multiple charts. E.g you install KFP, model registy and Notebooks. Then helm finding what is needed as dependencies (istio, knative etc.) and setting common thing is the big benefit. Otherwise i can just wrap kustomize as it is in helm and be done in a few minutes with GSOC.
